chore: improve "tree shaking" support and documentation

ncoden opened the pull request:
<!--- --------------------------------------------------------------------- -->
<!---                 Please fill the following template                    -->
<!---             Your pull request may be ignored otherwise                -->
<!--- --------------------------------------------------------------------- -->

## Description
<!--- Describe your changes in detail. Include any relevant information     -->
<!--- (reasons, difficulties, links to web references, related issues...)   -->

This pull request change some Foundation entries so they better follow the recommended practices and should be better analyzed by bundlers in order to enable tree-shaking. Also, it adds a section in the "Javascript" documentation with advices to reduce the final bundler size.

### Changes
* Add a simpler export file `js/foundation.js`
* Use `js/foundation.js` instead of `js/entries/foundation.js`
* Add default export as it may be sometimes required for tree-shaking
* Add "Tree Shaking" section in javascript documentation

### Others changes
* Add missing export of `Move` in "foundation-plugins" entry

<!--- Bugs and new features/improvements must be presented and discussed in -->
<!--- an issue first. Please create one if there is no issue related to     -->
<!--- this pull request.                                                    -->
- Related to https://github.com/zurb/foundation-sites/issues/11401

## Types of changes
<!--- What types of changes does your code introduce?                       -->
<!--- Put an `x` in all the boxes that apply:                               -->
- [x] Documentation
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
      functionality to change)

## Checklist (all required):
<!--- Go over all the following points, and put an `x` in the boxes.        -->
<!--- If you're unsure about any of these, don't hesitate to ask.           -->
- [x] I have read and follow the [CONTRIBUTING](CONTRIBUTING.md) document.
- [x] There are no other pull request similar to this one.
- [x] The pull request title is descriptive.
- [x] The template is fully and correctly filled.
- [x] The pull request targets the right branch (`develop` or `develop-v...`).
- [x] My commits are correctly titled and contain all relevant information.
- [x] My code follows the code style of this project.
- [x] I have updated the documentation accordingly to my changes (if relevant).
- ~~I have added tests to cover my changes (if relevant).~~
- [x] All new and existing tests passed.

<!--- --------------------------------------------------------------------- -->
<!---       For more information, see the CONTRIBUTING.md document          -->
<!---         Thank you for your pull request and happy coding ;)           -->
<!--- --------------------------------------------------------------------- -->

ncoden committed:
chore: use a more tree-shaking compliant ESM entry file

ncoden committed:
fix: add missing export of "Move" in "foundation-plugins" entry

ncoden committed:
chore: add default export as it may be sometimes required for tree-shaking

ncoden committed:
docs: add "Tree Shaking" section in javascript documentation

DanielRuf reviewed (commented):
Not sure if we really need two exports as only the default export is needed (afaik).

DaSchTour commented:
Well I think using default export is bad practice, because it allows doing very creepy things and also breaks checks for instanceof.
Here are just some other ideas why default export never were a good idea
https://blog.neufund.org/why-we-have-banned-default-exports-and-you-should-do-the-same-d51fdc2cf2ad
https://basarat.gitbooks.io/typescript/docs/tips/defaultIsBad.html

Another thing to consider is, that named exports also help with not breaking the API. If you have a default export every change to this export is possibly a breaking change.

So just never ever use default exports.

DanielRuf commented:
@DaSchTour I'm aware of this but it seems even Twitter Bootstrap (and many others) use `export default`, see https://github.com/twbs/bootstrap/blob/v4-dev/js/src/carousel.js#L520

ncoden commented:
It is not recommended to use default **imports**. But I remember that in some environnements they may be the only possible solution, and seems to work better with tree shaking. I'll try to find back the articles explaining this.

SassNinja commented:
@DaSchTour default export is not bad imo – if used correctly.

If someone misuses it to bundle several export in one default export it's of course bad practice and interferes with tree shaking. So you shouldn't do

**`utils.js`**
```javascript
function triple(val) {
  return val * 3;
}
function double(val) {
  return val * 2;
}
export default {
  triple: triple,
  double: double
};
```
**`app.js`**
```javascript
import utils from './utils';
console.log(utils.triple(10));
```

but instead do

**`utils.js`**
```javascript
export function triple(val) {
  return val * 3;
}
export function double(val) {
  return val * 2;
}
```
**`app.js`**
```javascript
import { triple } from './utils';
console.log(triple(10));
```

In contrast if you've a module that has just one export because it doesn't make sense to import partials of it then it's absolutely ok in my eyes to use export default. This happens if there's just one main function or class.

**`super-carousel.js`**
```javascript
class SuperCarousel {
  constructor() {
    console.log(`I'm a super carousel!`);
  }
}
export default SuperButton;
```
**`app.js`**
```javascript
import SuperCarousel from './super-carousel';
const carousel = new SuperCarousel();
```

SassNinja commented on docs/pages/javascript.md:
````diff
@@ -63,6 +63,19 @@ var Foundation = require('foundation-sites');
+Sadly, the "static analysis" promised by these bundlers to detect unused code in dependencies does not usually work, so we recommend you to manually import the Foundation plugins instead, like the following:
+
+```js 
+import Dropdown from 'foundation-sites/js/foundation.dropdown';
````
@ncoden I don't like this way of using foundation...

Is there a way to enable something more user friendly (while still being tree shakable)? Such as
```javascript
import { Reveal } from 'foundation-sites';
```

marvinhuebner replied:
@SassNinja i've also tested this pr (see [here](https://github.com/zurb/foundation-sites/issues/11401#issuecomment-412483242)). Would also prefer if there is a way when it's enough to do something like 
```js
import { Reveal } from 'foundation-sites';

$(document).foundation();
```
But it seems that this is not working, not with the current implementation.

Currently i have to do this to get it working: 
```js
import Foundation from 'foundation-sites/js/foundation';
import Reveal from 'foundation-sites/js/foundation.reveal';

var test = new Foundation.Reveal($('.reveal'));
```

DanielRuf replied:
Foundation needs more than just the plugin to work properly, if you mean that.

ncoden replied:
@SassNinja Sadly, as explained in the paragraph just above, the following _should work in theory_...
```js
import { Reveal } from 'foundation-sites';
```

...but **in practice** does not, as "pure modules" are not well detected by bundlers, and a direct import is recommended. For example the `lodash` library faced the same problem and they created [babel-plugin-lodash](https://github.com/lodash/babel-plugin-lodash) to convert things like `import map from 'lodash'` to `import _map from 'lodash/map'`.

So at best it depends on the user stack to choose whether he can use the named import and try to keep tree shaking working.

ncoden replied:
Poke @jdalton, It would be nice to get your opinon on this 😉.

ncoden replied:
@marvinhuebner I prefer the `import { Reveal } from 'foundation-sites';` approach too. But I just want to make something that works 😉

jdalton replied:
For side-effect free ability: `lodash-es` is detected as pure by webpack v4 and parcel.js with the `lodash-es`'s use of the package.json field `"sideEffects": false,`.

ncoden replied:
I tested with `"sideEffects": false,` and followed all the ressources I found on tree shaking. I am unable to make this work with imports from `foundation-sites`.

Take a look at https://github.com/ncoden/foundation-sites-tree-shaking. Please tell me if I got something wrong. The project rely on this branch `chore/tree-shaking`, but locally it's linked to my local `foundation-sites` clone.

jdalton replied:
`sideEffects: false` is a per package setting not an application setting. This is why it's set in `lodash-es` because I, as the package author, am stating that its modules are side-effect free.

ncoden replied:
> `sideEffects: false` is a per package setting

Yes I know. I set `sideEffects: false` locally in `foundation-sites`

> but locally it's linked to my local foundation-sites clone with sideEffects set.

DaSchTour replied:
I'm curious. As Foundation uses jQuery and extends it and also relies on extending global Foundation object I'm not really sure if Foundation can be considered sideEffects free.

ncoden replied:
Foundation plugins can be considered as without side effects, as they must be imported and initialized on a component to work. It's the same for the Foundation Core that is manually added to jQuery `addTojQuery()`. I guess the "no side effects" only means: do not rely on others components without importing them. But Foundation core and plugins import all their dependencies.

ncoden replied:
@jdalton @DaSchTour Would you know what I am missing here (https://github.com/ncoden/foundation-sites-tree-shaking) ? I don't see what I could configure make tree shaking working with imports from `foundation-sites`.

ncoden replied:
@SassNinja @DaSchTour  So what's your opinon on this ?

DaSchTour replied:
@ncoden I have no idea. I never had any problems with tree shaking but I also haven't used or dependend on any globals for a long time.

ncoden committed:
chore: remove unrecommended "grouped" default exports

Default exports should only be used for "primary" objects or classes, not for libraries for which individual named exports should be prefered.

ncoden commented:
@SassNinja I removed "grouped" named exports:

> Default exports should only be used for "primary" objects or classes, not for libraries for which individual named exports should be prefered.

Also, I kept named exports at the end of the files. See https://github.com/airbnb/javascript/issues/1275#issuecomment-274367523

ncoden commented:
@DaSchTour IMHO default imports are not a good import practice, but it is not a bad export one. There is many situation where default imports is the only solution, like when trying to simply transition from ES5 to ES6 (so from CommonJS to ESM modules) or when module imports are handled and/or transpiled by the framework's CLI (like for Ember).

DaSchTour commented:
@ncoden I know this is a constructed scenario. But one thing that can happen with default exports, for example because of somebody making a mistake while resolving merge conflicts, you end up with code like this
```js
import Reveal from 'foundation-sites/js/foundation.dropdown';
import Dropdown from 'foundation-sites/js/foundation.reveal';
```
This is not an error and with JS you might search for hours to find the reason why everything is messed up.

ncoden commented:
@DaSchTour Yes. Default imports should be avoided when possible for this reason. But we should still propose default import for the reasons given above, even if we do not _encourage_ it.

DaSchTour commented:
@ncoden I don't see any reason why default imports should be possible. In general I wonder why put so much effort into making foundation 6 tree shakable or why put any effort into foundation 6 at all while everybody is waiting for 7.

ncoden commented:
> why put so much effort into making foundation 6 tree shakable

Because fixing and cleaning the ESM+AMD+CJS+static imports (which is a big "feature" of the v6.5.0 release) goes with providing efficient ways to use them in each of these environnement. And IMHO this means allowing to import single plugins without importing the whole framework (which is unexpected when you have the ability to cherry-pick feature).

>  why put any effort into foundation 6 at all while everybody is waiting for 7

Good question. Personally, I'd like to get Foundation 6 _stable_ before we take 6-12  months to prepare and work on Foundation 7. But I recognize that this step took way much time than expected. I am sorry about that.

SassNinja commented:
> Good question. Personally, I'd like to get Foundation 6 stable before we take 6-12 months to prepare and work on Foundation 7. But I recognize that this step took way much time than expected. I am sorry about that. – @ncoden 

I don't think it's wrong to enable using Foundation v6 with module bundler the way users expect (import only certain plugins).

It may seem a bit waste of time regarding v7 but I think it's more important to ensure a good usability of the existing version. If users get pissed of using Foundation they may not wait until v7...

DaSchTour commented:
@SassNinja well I guess some users already won't be waiting for v7. It's called Osborne effect.

ncoden commented:
@DaSchTour @SassNinja Could we not enter into meta-discussions about the global Foundation roadmap and focus on the current issue ? Time matters, everyone agree on that, but we're not saving any with this kind of discussion. For now I'd like to get this PR approved/rejected to release RC3.

ncoden committed:
chore: add missing "sideEffects" flag to ensure tree shaking with webpack

ncoden committed:
docs: make "direct import" for tree shaking sounds less imperative

ncoden commented:
@SassNinja 

I adapted the message so direct imports are only recommended if tree-shaking doesn't work, "depending on your build environment". Are you good with this ?

@jdalton 

I added the `sideEffects` flag, I think this is the best I can do.

SassNinja commented:
> I adapted the message so direct imports are only recommended if tree-shaking doesn't work, "depending on your build environment". Are you good with this ? – @ncoden 

Not really :/

But I've now set up a simple webpack config using your feature branch and tree shaking did work without problems. The only downside is all plugins still appear in the stats but the emitted JS code looks good.

```javascript
// webpack.example.js

const path = require('path');

module.exports = {
    mode: 'production',
    devtool: 'none',
    entry: {
        'example': './example.js'
    },
    output: {
        path: path.resolve(__dirname, 'dist-example'),
        filename: '[name].js'
    },
    externals: {
        jquery: 'jQuery'
    },
    optimization: {
        minimize: false,
        runtimeChunk: 'single'
    }
};
```

```javascript
// example.js

import { Abide } from './js/foundation.js';

console.log(typeof Abide);
```

How did your setup look where tree shaking isn't working?

ncoden commented:
@SassNinja See https://github.com/ncoden/foundation-sites-tree-shaking

ncoden commented:
@SassNinja I cannot make tree shaking work with your configuration. Could you share your dummy project in a repo ?

SassNinja commented:
> I cannot make tree shaking working with your webpack configuration. Could you share your dummy project in a repo ?

Here's an example repo:
https://github.com/SassNinja/example-foundation-sites-tree-shaking

I think I know what went wrong on your side because I also faced an issue when creating that example repo. To work the bundler needs to take the `module` field in the `package.json` what webpack does not do by default.

So you need to override it like this
```javascript
resolve: {
    mainFields: ['module', 'browser', 'main']
}
```

SassNinja commented:
> Thank you. Too bad that the webpack documentation about tree shaking doesn't mention it.

Maybe we should adjust our package.json to fix this without the need for additional bundler config. The `browser` field looks promising to me.

See https://github.com/webpack/webpack/issues/4674#issuecomment-355853969

What do you think?

SassNinja commented:
I've simply replaced the field name `module` with `browser`.
Afterwards it's working for me without further bundler config (tested webpack).

SassNinja commented:
A radical idea: what about removing the whole `entries` folder?

I think if you use a module bundler (and therefore probably want tree shake) you have to include the desired plugins via
```javascript
import { Abide, OffCanvas } from 'foundation-sites';
```
We can provide a list with all plugins imports for copy & paste.

And people who wanna import the whole framework without a module bundler (e.g. when using gulp without any bundler in the task) may use the dist file directly.

We could also cover the dist case with an export default but I don't see any useful reason to offer something like

```javascript
import FoundationEverything from '../dist/js/foundation.js';
export default FoundationEverything;
```

ncoden commented:
> I've simply replaced the field name module with browser.
> Afterwards it's working for me without further bundler config (tested webpack).

Thank you, I'll test this tonigh and update the pull request.

> A radical idea: what about removing the whole `entries` folder?

Hmmm, this would break backward compatibility so it's definitely not for this pull request. For v6.6.0, maybe if it enables anything useful. But for now I don't see any advantages for us or our users.

ncoden commented:
I removed the `browser` field it and it works well. According to the [specs](https://github.com/defunctzombie/package-browser-field-spec) there is no reason for us to use a `browser` field as the package itself it only for browsers (or any environement with a DOM). With this configuration, ESM-compliant transpilers/bundlers will pick `module`, and the others will pick `main`.

There is still one problem too: the "picked" code is not only a ESM bundle, but also ES6 code. Webpack will use `module` because it's an ESM bundle but will not transpile the ES6 stuff...

So I think we should also provide our distribution files in the `.esm.js` format: a transpiled file BUT with ESM exports untouched, and use it for the `module` field.

See for example the [Vue dist folder](https://github.com/vuejs/vue/tree/dev/dist): `vue.js` (UMD), `vue.common.js` (commonJs) and `vue.esm.js` (ESM)

What do you think ?

SassNinja commented:
> According to the specs there is no reason for us to use a browser field as the package itself it only for browsers (or any environement with a DOM).

Alright then let's ditch `browser` and only use `main` and `module`

> Webpack will use module because it's an ESM bundle but will not transpile the ES6 stuff...

I don't understand this – what do you mean?

The code gets picked up by the bundler which understands ES6 and resolves all dependencies. If necessary it can be transpiled to ES5 using e.g. the babel-loader.

> See for example the Vue dist folder: vue.js (UMD), vue.common.js (commonJs) and vue.esm.js (ESM)

I think a transpiled ESM file is nice to speed the build process up as there's no need then to transpile node_modules yourself (while preserving tree shake support).
However we should also offer an untouched file imo (see `vue.esm.browser.js`) for people who don't need any pre transpilation.

ncoden commented:
> However we should also offer an untouched file imo (see vue.esm.browser.js) for people who don't need any pre transpilation.

You are right. I'm going for the following: `foundation.js` (UMD), `founcation.cjs.js` (commonJs), `foundation.esm.js` (transpiled ESM module), `foundation.es6.js` (untouched ESM module).

DaSchTour commented:
What is the difference between transpiled ESM module and untouched ESM module? Transpiling ESM module to let's say ES5 would make it "remove" modules because ES5 doesn't have modules. And I think any kind of processing might possibly hurt treeshaking and also minification.

SassNinja commented:
> What is the difference between transpiled ESM module and untouched ESM module? Transpiling ESM module to let's say ES5 would make it "remove" modules because ES5 doesn't have modules. – @DaSchTour 

That's why I wrote `foundation.esm.js` would still contain ES6 code, precisely the import/export stuff.
But compared to `foundation.es6.js` (@ncoden maybe we should rename it to avoid confusion?) it's compiled to ES5 except of all import/export.

> And I think any kind of processing might possibly hurt treeshaking – @DaSchTour 

afaik webpack requires ES6 import/export for tree shaking. If we preserve that on transpilation it shouldn't be a problem and improves the build time because users can exclude foundation from their own ES5 transpilation then.
Anyway this has to be tested of course (haven't done this yet).

ncoden commented:
Some bundlers (webpack, rollup...) supports the ES6 module format and require to use it for tree shaking but will leave the other ES6 stuff intact (they are bundlers, not transpilers). `foundation.esm.js` is transpiled ES5 code embeded in ESM modules for those bundlers. `foundation.es6.js` is untouched ES6 code and modules that can be manually transpiled by our users depending on their targets.

ncoden committed:
chore: remove redundant "pkg.browser" field

According to the specification for `pkg.browser`, there is no reason for us to use this field as the package itself it only for browsers (or any environement with a DOM). Without the field, ESM-compliant transpilers/bundlers will pick `module`, and the others will pick `main`.

https://github.com/defunctzombie/package-browser-field-spec

ncoden committed:
docs: improve "tree shaking" documentation to promote named imports

ncoden commented:
@SassNinja So I made (a lot of) tests:
- ESM6+ bundles (es6+ in es modules) are correctly shaked by both Webpack and Rollup.
- ESM5 bundles (transpilled es5 in es modules) are not shaked, despites being marked as `__PURE__` by Rollup. This comes from static properties defined outside of the class like `<plugin>.defaults = { ... }`, and plugin initializations in entry files.

There is no other way to define static properties for now. There is a [TC39 proposal](https://github.com/tc39/proposal-class-fields) for this but the [Babel plugin](https://babeljs.io/docs/en/next/babel-plugin-proposal-class-properties.html) generate code outside of the class closure, making it [incompatible with tree sharking](https://github.com/babel/babel/issues/5902).

For now I think we can go with only an ES6+ version directly from the sources (so Webpack/Rollup users will have to transpile the code), and I'll do some refactor later to allow us to provide shakable transpiled bundles.

I removed `pkg.browser`, and improved the "tree shaking" section in the installation documentation to promote named imports over "manual" imports. What do you think ?

SassNinja commented:
> There is no other way to define static properties for now. There is a TC39 proposal for this but the Babel plugin generate code outside of the class closure, making it incompatible with tree sharking.
>
> For now I think we can stay with the ES6 sources in pkg.modules (so Webpack/Rollup users will have to transpile the code), and I'll do some refactor later to allow us to provide shakable transpiled bundles.

Ok, then let's stay with untranspiled ES6 code in the module field.
Would be of course much better if users don't need to transpile node modules themselves but does not make sense if tree shaking doesn't work correctly anymore.

> I removed pkg.browser, and improved the "tree shaking" section in the installation documentation to promote named imports over "manual" imports. What do you think ?

Docs lgtm :+1: 

The only thing I still doesn't like is the usage of named **and** default export in the plugins that export the same.
I've only tested webpack and there it also works without default export.
But you've also tested other bundlers and noticed it's necessary to have default exports, haven't you?

ncoden commented:
@SassNinja Thank you for the review :)

> But you've also tested other bundlers and noticed it's necessary to have default exports, haven't you?

I remember that yes, but I'll check again just to be sure.

ncoden committed:
Revert "chore: add default export as it may be sometimes required for tree-shaking"

This reverts commit 5d9373b3dce67e29769a7afd73598dc609dc7dc1.

ncoden commented:
I tested, and issues actually come fron side-effects in the "fully-initialized" browser entry `js/entries/foundation.js`. But we can't move it from there as this is how Foundation is used: `import { Foundation } from "foundation-sites"; // already initialized`
